src/Middleware/WebSockets/src/HandshakeHelpers.cs

```diff
@@ -72,5 +75,89 @@ public static string CreateResponseKey(string requestKey)
 
             return Convert.ToBase64String(hashedBytes);
         }
+
+        public static bool ParseDeflateOptions(ReadOnlySpan<char> extension, WebSocketDeflateOptions options, [NotNullWhen(true)] out string? response)
+        {
+            response = null;
+            var builder = new StringBuilder(WebSocketDeflateConstants.MaxExtensionLength);
+            builder.Append(WebSocketDeflateConstants.Extension);
+
+            while (true)
+            {
+                int end = extension.IndexOf(';');
+                ReadOnlySpan<char> value = (end >= 0 ? extension[..end] : extension).Trim();
+
+                if (value.Length > 0)
+                {
+                    if (value.SequenceEqual(WebSocketDeflateConstants.ClientNoContextTakeover))
+                    {
+                        // REVIEW: If someone specifies true for server options, do we allow client to override this?
+                        options.ClientContextTakeover = false;
+                        builder.Append("; ").Append(WebSocketDeflateConstants.ClientNoContextTakeover);
+                    }
+                    else if (value.SequenceEqual(WebSocketDeflateConstants.ServerNoContextTakeover))
+                    {
+                        options.ServerContextTakeover = false;
+                        builder.Append("; ").Append(WebSocketDeflateConstants.ServerNoContextTakeover);
+                    }
+                    else if (value.StartsWith(WebSocketDeflateConstants.ClientMaxWindowBits))
+                    {
+                        var clientMaxWindowBits = ParseWindowBits(value, WebSocketDeflateConstants.ClientMaxWindowBits);
+                        if (clientMaxWindowBits > options.ClientMaxWindowBits)
+                        {
+                            return false;
+                        }
+                        // if client didn't send a value for ClientMaxWindowBits use the value the server set
+                        options.ClientMaxWindowBits = clientMaxWindowBits ?? options.ClientMaxWindowBits;
+                        builder.Append("; ").Append(WebSocketDeflateConstants.ClientMaxWindowBits).Append('=')
+                            .Append(options.ClientMaxWindowBits.ToString(CultureInfo.InvariantCulture));
+                    }
+                    else if (value.StartsWith(WebSocketDeflateConstants.ServerMaxWindowBits))
```

Comment:
The server still needs to send back these limits even if it doesn't receive it from the client.